ocaml/xapi/xapi_vm_group_helpers.ml

```diff
@@ -0,0 +1,173 @@
+           Db.VM.get_power_state ~__context ~self:vm = `Running
+       )
+  in
+  let get_host vm =
+    let host =
+      match Db.VM.get_resident_on ~__context ~self:vm with
+      | h when h <> Ref.null ->
+          Some h
+      | _ ->
+          None
+    in
+    host
+  in
+  match vms with
+  | [] | [_] ->
+      None
+  | vm :: _ ->
+      let host = get_host vm in
+      let same_host = List.for_all (fun vm' -> get_host vm' = host) vms in
+      debug "The breach state of group (UUID: %s) is :%b"
+        (Db.VM_group.get_uuid ~__context ~self:group)
+        same_host ;
+      if same_host then host else None
+
+let report_anti_affinity_alert ~__context ~group ~host =
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  let host_uuid = Db.Host.get_uuid ~__context ~self:host in
+  let body =
+    Printf.sprintf
+      "<body><message>Breach on VM anti-affinity rules</message><VM_group>"
+    ^ group_uuid
+    ^ "</VM_group><host>"
+    ^ host_uuid
+    ^ "</host></body>"
+  in
+  let obj_uuid =
+    Db.Pool.get_uuid ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  Xapi_alert.add ~msg:Api_messages.breach_on_vm_anti_affinity_rules ~cls:`Pool
+    ~obj_uuid ~body
+
+let get_all_records ~__context =
+  let records =
+    Helpers.call_api_functions ~__context (fun rpc session_id ->
+        Client.Client.Message.get_all_records ~rpc ~session_id
+    )
+  in
+  records
+
+let filter_alert_with_group ~__context ~group ~records =
+  let contains_substring str substr =
+    let regexp = Str.regexp_string substr in
+    try
+      ignore (Str.search_forward regexp str 0) ;
+      true
+    with Not_found -> false
+  in
+  let group_uuid = Db.VM_group.get_uuid ~__context ~self:group in
+  records
+  |> List.filter (fun (_, record) ->
+         record.API.message_name
+         = fst Api_messages.breach_on_vm_anti_affinity_rules
+         && contains_substring record.API.message_body group_uuid
+     )
+
+(** If it is breach and no alerts existed, generate one, 
+    If it is not breach and alerts existed, dismiss the existed alert *)
+let update_vm_placement_alert_with_group ~__context ~group ~records =
+  try
+    let breach_on_host = check_breach ~__context ~group in
+    let alert_records = filter_alert_with_group ~__context ~group ~records in
+    debug "The length of group (UUID: %s) is: %d"
+      (Db.VM_group.get_uuid ~__context ~self:group)
+      (List.length alert_records) ;
+    match (alert_records, breach_on_host) with
+    | [], Some host ->
+        report_anti_affinity_alert ~__context ~group ~host
+    | records, None ->
+        List.iter
+          (fun (ref, _) ->
+            Helpers.call_api_functions ~__context (fun rpc session_id ->
+                Client.Client.Message.destroy ~rpc ~session_id ~self:ref
+            )
+          )
+          records
+    | _, _ ->
+        ()
+  with _ -> warn "Update VM anti-affinity alert failed"
+
+let update_vm_placement_alert_with_vm ~__context ~vm =
```

Comment:
```
let maybe_update_alerts_for_vm ~__context ~vm =
```

Reply:
How about `maybe_update_vm_anti_affinity_alert_for_vm`

Reply:
And it could be:
```
let maybe_update_alerts_for_vm ~__context ~vm =
  Db.VM.get_groups ~__context ~self:vm
  |> List.filter (fun g ->
      Db.VM_group.get_placement ~__context ~self:g  = `anti_affinity
  )
  |> function
  | [] ->
      ()
  | group :: _ ->
      let alerts = alerts_of_group ~__context ~group in
      update_alerts_of_group ~__context ~group:g ~alerts
```